src/entities/pools/weighted/weightedJoin.ts

```diff
@@ -0,0 +1,226 @@
+import { createPublicClient, encodeFunctionData, http } from 'viem';
+import {
+    BaseJoin,
+    JoinCallInput,
+    JoinInput,
+    JoinQueryResult,
+    PoolState,
+} from '..';
+import { Address } from '../../../types';
+import {
+    BALANCER_HELPERS,
+    BALANCER_VAULT,
+    CHAINS,
+    ZERO_ADDRESS,
+} from '../../../utils';
+import { WeightedEncoder } from '../../encoders/weighted';
+import { TokenAmount } from '../../tokenAmount';
+import { balancerHelpersAbi } from '../../../abi/balancerHelpers';
+import { Token } from '../../token';
+import { vaultAbi } from '../../../abi';
+
+export class WeightedJoin implements BaseJoin {
+    // TODO - Probably not needed
+    getInstance(): WeightedJoin {
+        return new WeightedJoin();
+    }
+
+    public async query(
+        input: JoinInput,
```

Comment:
I get the goal here, but in trying to make a one size fits all, it seems like you're creating an interface that is quite challenging to reason about. For given out, I need to know that the `input.tokenAmounts.length === 1` and that it should only contain the bpt. The only way I would be able to figure that out is by reading the docs or reading the code, I don't think I would ever intuitively know that.

Reply:
I think that a good approach to fix this would be John's initial draft, where he provided different input types, that make it easier for the user to make sense of all possible join variants.

Since it's a small change, I'll go ahead and update the PR, so we can follow-up the discussion with more concrete implementation.